File: leetcode/97_interleaving_string.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int n1 = s1.size();
        int n2 = s2.size();
        int n3 = s3.size();
        if (n1 + n2 != n3) {
            return false;
        }
        vector<vector<int>> memo(n1, vector<int> (n2, -1));
        return isInterLeaveRecursive(s1, s2, s3, 0, 0, 0, memo);
    }
    bool isInterLeaveRecursive(string s1, string s2, string s3, int t1, int t2, int t3, vector<vector<int>>& memo) {
        int n1 = s1.size();
        int n2 = s2.size();
        int n3 = s3.size();
        if (t1 < n1 && t2 < n2 && memo[t1][t2] > -1) {
            return !!memo[t1][t2];
        }
        if (t3 >= n3) {
            return true;
        }
        bool result = false;
        if (t1 < n1 && s1[t1] == s3[t3]) {
            result |= isInterLeaveRecursive(s1, s2, s3, t1+1, t2, t3+1, memo);
        }
        if (t2 < n2 && s2[t2] == s3[t3]) {
            result |= isInterLeaveRecursive(s1, s2, s3, t1, t2+1, t3+1, memo);
        }
        if (t1 < n1 && t2 < n2) {
            memo[t1][t2] = (int)result;
        }
        return result;
    }
};

File: leetcode/97_interleaving_string.py
class Solution:
    # @param A : string
    # @param B : string
    # @param C : string
    # @return an integer
    def isInterleave(self, A, B, C):
        n1  = len(A)
        n2 = len(B)
        n3 = len(C)
        if n3 != n1 + n2:
            return False
        def isInterleaveRecursive(index1, index2, index3):
            if index3 >= n3:
                return True
            result = False
            if(index1 < n1 and A[index1] == C[index3]):
                result |= isInterleaveRecursive(index1+1, index2, index3+1)
            if(index2 < n2 and B[index2] == C[index3]):
                result |= isInterleaveRecursive(index1, index2+1, index3+1)
            return result
        return isInterleaveRecursive(0, 0, 0)
